openrtb/banner_test.go

```diff
@@ -0,0 +1,105 @@
+				BlockedAttributes: []int{2},
+				Position:          openrtb.AdPositionAboveFold,
+				TopFrame:          &testInt,
+				ExpandDirections:  []int{3},
+				APIFrameworks:     []int{4},
+				Width:             1,
+				Height:            1,
+			},
+		},
+	}
+
+	for _, testCase := range testCases {
+		bannerCopy := testCase.banner.Copy()
+
+		if testCase.banner.ID != nil {
+			if &testCase.banner.ID == &bannerCopy.ID {
+				t.Errorf("Address of ID should not be the same in copied banner. ID1: %p ID2: %p.", testCase.banner.ID, bannerCopy.ID)
+			}
+		}
+
+		if testCase.banner.BlockedTypes != nil {
+			if &testCase.banner.BlockedTypes == &bannerCopy.BlockedTypes {
+				t.Errorf("Address of BlockedTypes should not be the same in copied banner. BlockedTypes1: %p BlockedTypes2: %p.", testCase.banner.BlockedTypes, bannerCopy.BlockedTypes)
+			}
+		}
+
+		if testCase.banner.BlockedAttributes != nil {
+			if &testCase.banner.BlockedAttributes == &bannerCopy.BlockedAttributes {
+				t.Errorf("Address of BlockedAttributes should not be the same in copied banner. BlockedAttributes1: %p BlockedAttributes2: %p.", testCase.banner.BlockedAttributes, bannerCopy.BlockedAttributes)
+			}
+		}
+
+		if testCase.banner.MIMETypes != nil {
+			if &testCase.banner.MIMETypes == &bannerCopy.MIMETypes {
+				t.Errorf("Address of MIMETypes should not be the same in copied banner. MIMETypes1: %p MIMETypes2: %p.", testCase.banner.MIMETypes, bannerCopy.MIMETypes)
+			}
+		}
+
+		if testCase.banner.TopFrame != nil {
+			if &testCase.banner.TopFrame == &bannerCopy.TopFrame {
+				t.Errorf("Address of TopFrame should not be the same in copied banner. TopFrame1: %p TopFrame2: %p.", testCase.banner.TopFrame, bannerCopy.TopFrame)
+			}
+		}
+
+		if testCase.banner.ExpandDirections != nil {
+			if &testCase.banner.ExpandDirections == &bannerCopy.ExpandDirections {
+				t.Errorf("Address of ExpandDirections should not be the same in copied banner. ExpandDirections1: %p ExpandDirections2: %p.", testCase.banner.ExpandDirections, bannerCopy.ExpandDirections)
+			}
+		}
+
+		if testCase.banner.APIFrameworks != nil {
+			if &testCase.banner.APIFrameworks == &bannerCopy.APIFrameworks {
+				t.Errorf("Address of APIFrameworks should not be the same in copied banner. APIFrameworks1: %p APIFrameworks2: %p.", testCase.banner.APIFrameworks, bannerCopy.APIFrameworks)
+			}
+		}
+
+		if testCase.banner.ID != nil {
+			if &testCase.banner.ID == &bannerCopy.ID {
+				t.Errorf("Address of ID should not be the same in copied banner. ID1: %p ID2: %p.", testCase.banner.ID, bannerCopy.ID)
+			}
+		}
+
+		if !reflect.DeepEqual(testCase.banner, bannerCopy) {
+			banner1JSON, _ := json.MarshalIndent(testCase.banner, "", "  ")
+			banner2JSON, _ := json.MarshalIndent(bannerCopy, "", "  ")
+			t.Errorf("Banners should hold the same values.\nExpected: %s\n Got: %s", banner1JSON, banner2JSON)
+		}
+
+		if diff := deep.Equal(testCase.banner, bannerCopy); diff != nil {
+			t.Error(diff)
```

Comment:
Missing description here.

Reply:
The Deep library could output the different.